R/slide.R

```diff
@@ -318,36 +336,80 @@ epi_slide <- function(
       .data_group <- filter(.data_group, o)
     }
 
-    result <-
-      if (is.null(.new_col_name)) {
-        if (inherits(slide_values, "data.frame")) {
-          # unpack into separate columns (without name prefix) and, if there are
-          # re-bindings, make the last one win for determining column value &
-          # column placement:
-          mutate(.data_group, slide_values)
-        } else {
-          # apply default name:
-          mutate(.data_group, slide_value = slide_values)
+    # To label the result, we will parallel some code from `epix_slide`, though
+    # some logic is different and some optimizations are less likely to be
+    # needed as we're at a different loop depth.
+
+    # Unlike `epix_slide`, we will not every have to deal with a 0-row
+    # `.group_key`: we return early if `epi_slide`'s `.x` has 0 rows, and our
+    # loop over groups is the outer loop (>= 1 row into the group loop ensures
+    # we will have only 1-row `.group_key`s). Further, unlike `epix_slide`, we
+    # actually will be using `.group_data` rather than work with `.group_key` at
+    # all, in order to keep the pre-existing non-key columns. We will also try
+    # to work directly with `epi_df`s instead of listified tibbles; since we're
+    # not in as tight of a loop, the increased overhead hopefully won't matter.
+    # We'll need to use `bind_cols` rather than `c` to avoid losing
+    # `epi_df`ness.
+
+    res <- .data_group
+
+    if (is.null(.group_new_col_name)) {
+      if (inherits(slide_values, "data.frame")) {
+        # Sometimes slide_values can parrot back columns already in `res`; allow
+        # this, but balk if a column has the same name as one in `res` but a
+        # different value:
+        comp_nms <- names(slide_values)
+        overlaps_existing_names <- comp_nms %in% names(res)
+        for (comp_i in which(overlaps_existing_names)) {
+          if (!identical(slide_values[[comp_i]], res[[comp_nms[[comp_i]]]])) {
```

Comment:
The whole point of the name of this PR is to allow this, though.  So that user can do
`%>% group_by(geo_value) %>% epi[x]_slide(~ forecaster(.x), .....)`
on a `forecaster` expecting `.x` to be an `epi_df`.
- To deliver `forecaster` an `epi_df`, we need to include the `geo_value` column.
- If a `forecaster` gets something with a `geo_value` column, it should probably output a `geo_value` column as well.
- Now we have a duplicate `geo_value` column.

If the `identical` check is too expensive, then we could consider some alternatives, but I think this is probably the user-friendliest without doing some complex epiprocess changes + updating epipredict.  I'm hoping it's not actually too expensive, as:
- `epi_slide` only checks this once per group
- `epix_slide` checks this once per user comp, but user comps are generally beefier there

So.... let's benchmark first?

Reply:
Potential optimizations: deduping vs. `.group_key` can probably be a bit faster using all(<user col> == <scalar>).  Other labels will need to use `identical` or `all(<vec> == <vec>)`, assuming we can rely on `==`....  I've tested some other approaches backed by `unique` to dedupe but they seem slower.

todo: think about factor column correctness

Reply:
Thanks for the context, I didn't know what functionality you were trying to add in this PR or if you were just bug fixing. 

I have the hunch that the `geo_value` vector of `.data_group` and the `geo_value` vector of `forecaster(.x)` are going to have annoyingly differently lengths sometimes. For instance, if the forecaster returns multiple aheads. That's going to lead to a mismatch here right? 

I feel like the user has a pretty simple workaround here of `~forecaster(.x)` to `function(x) {forecaster(x) %>% select(-geo_value)}` for any collision issues they get.

Reply:
Also, how does the user specify that their `.f` takes an epi_df rather than a vector? Isn't it always a vector as given by slider::hop_index?

Reply:
> I have the hunch that the geo_value vector of .data_group and the geo_value vector of forecaster(.x) are going to have annoyingly differently lengths sometimes. For instance, if the forecaster returns multiple aheads. That's going to lead to a mismatch here right?

For `epi_slide`, yes, but they would have a mismatch anyway regardless of whether there was a `geo_value` column, since their other columns would also be a mismatch to the required number of rows.

For `epix_slide`, no, or at least it shouldn't.  We should be rep-ing the .group_key to the same size as the user's computation, so the geo_value columns should be the same length when checking.  That's not the most efficient way though, so I can optimize that a little.

> I feel like the user has a pretty simple workaround here of ~forecaster(.x) to function(x) {forecaster(x) %>% select(-geo_value)} for any collision issues they get. I don't know, this feels like a lot of very involved checks for an edge use case? Did this come up in our interns code?

Yes, it came up with them very early on, when they started trying to backtest things, both with epi_slide and epix_slide.  I helped them avoid for whatever they were doing; I'm not sure how long it'd take them to realize the workarounds required.  I think we just shouldn't require that workaround at all; epiprocess & epipredict should play nicely together, and this actually seems the simpler approach I can think of for making that happen.  (Though even with dedupe they still won't play nice in one scenario: epi_slide epikey-grouped single-ahead naive forecast backtesting with custom epi_recipes that output a different idea of the time_value; this apparently can output a time_value column not matching the .ref_time_value, and dedupe will complain about that mismatch, though that's probably better than overwriting one with the other.)

> Also, how does the user specify that their .f takes an epi_df rather than a vector? Isn't it always a vector as given by slider::hop_index?

`.x` has always been a tibble, `epi_df`, or `epi_archive`, depending on inputs.  This PR would standardize that a bit to make it an `epi_df` unless it's `epix_slide` with `.all_versions = TRUE`, `epi_archive` in that other case.

Reply:
Oh right, `select(-geo_value)` workaround only works if you have part of this PR that keeps `.x` as an `epi_df` with the `geo_value` attached (when you are geo-grouping); otherwise, you also need to reintroduce the geo_value column & convert to `epi_df` (& not forget `as_of = .ref_time_value` in some cases) for current epipredict to work, though I'm not sure if Dan's PR fully would fix this.  So fully workaround is a little more complicated.

Reply:
Performance note: we don't actually require any `identical` calls if they don't output any column name conflicts.  I'm hoping that simply checking for name conflicts # groups times in `epi_slide` and # comps times in `epix_slide` isn't going to be a big hit.

Reply:
> I'm hoping in those cases that simply checking for name conflicts # groups times in epi_slide and # comps times in epix_slide isn't going to be a big hit. 

That seems reasonable.

I think mostly I'm just reacting to the increase in complexity in epi_slide and I'm coming into this PR blind as to what issues you were aiming to address. I think I understand the issues here better now and it seems more reasonable to do these checks now. I don't want to derail this PR, so overall I think that the user experience looks more sane here, so thank you for the examples with workarounds above. In another round of changes, maybe we can explore handling these collections in group_modify.epi_df instead, but let's leave the conversation here for now.

Reply:
I've investigated all the potential changes above, and:
- Decided against just erroring out without attempting dedupe on some columns.  Seems like attempting dedupe on all overlapping columns might be helpful sometimes, and less complex code-wise.
- The more efficient scalar label vs. vector comp output comparison actually seems slower when you have appropriate type checks & generality.

Unless benchmarking guides us back to this area, I'm won't revisit these.  Since we're already benchmarking somewhere, I'll close this convo.